### PR TITLE
Mega Evolution Quest Filters

### PR DESCRIPTION
**Pull request by versx**
Fix: #186 
image

### CODE REVIEW

File: src/data/map.js [hunk 1]
```diff
@@ -1270,14 +1291,23 @@ const getAvailableRaidBosses = async () => {
...
     const itemResults = await db.query(sql);
     sql = 'SELECT quest_pokemon_id AS id FROM pokestop WHERE quest_reward_type=7 GROUP BY quest_pokemon_id';
     const pokemonResults = await db.query(sql);
+    sql = `
+    SELECT
+        quest_pokemon_id AS id,
+        json_extract(json_extract(quest_rewards, '$[*].info.amount'), '$[0]') AS amount
+    FROM rdmdb.pokestop
+    WHERE quest_reward_type = 12
+    GROUP BY quest_pokemon_id
```
Comment (clburlison) [inline src/data/map.js:1304]:
```suggestion
    SELECT
        DISTINCT json_extract(json_extract(quest_rewards, '$[*].info.pokemon_id'), '$[0]') AS id
    FROM pokestop
    WHERE quest_reward_type = 12
    GROUP BY id;
```

File: src/data/map.js [hunk 2]
```diff
@@ -481,6 +486,22 @@ const getPokestops = async (minLat, maxLat, minLon, maxLon, updated = 0, showPok
             excludePokemonSQL = sqlExcludeCreate;
         }
 
+        if (excludedEvolutions.length === 0) {
+            excludeEvolutionSQL = 'OR (quest_reward_type IS NOT NULL AND quest_reward_type = 12 AND quest_pokemon_id IS NOT NULL)';
+        } else {
+            let sqlExcludeCreate = 'OR (quest_reward_type IS NOT NULL AND quest_reward_type = 12 AND quest_pokemon_id IS NOT NULL AND quest_pokemon_id NOT IN (';
```
Comment (clburlison) [inline src/data/map.js:492]:
```suggestion
        if (excludedEvolutions.length === 0) {
            excludeEvolutionSQL = 'OR (quest_reward_type IS NOT NULL AND quest_reward_type = 12)';
        } else {
            let sqlExcludeCreate = 'OR (quest_reward_type IS NOT NULL AND quest_reward_type = 12 AND json_extract(json_extract(quest_rewards, "$[*].info.pokemon_id"), "$[0]") NOT IN (';
```

The above suggestion looks correct for me but doesn't seem to work...Not sure what I'm missing.

**Review by clburlison — CHANGES_REQUESTED** [review]
Are you using an old version of RDM? `quest_pokemon_id` should be null for these mega energy quests. Instead we need to json _extract from the `quest_rewards` column.

I made a few changes and got some progress but the exclude query wasn't working for me. Might just need 💤

File: src/data/map.js (hunk 1 above)
Comment (versx) [reply src/data/map.js:1304]: With `DISTINCT` I think we can drop the `GROUP BY id` clause.

**Comment by versx** [comment]
> Are you using an old version of RDM? `quest_pokemon_id` should be null for these mega energy quests. Instead we need to json _extract from the `quest_rewards` column.
> 
> I made a few changes and got some progress but the exclude query wasn't working for me. Might just need 💤

Is it a different key now? In my table it shows `quest_pokemon_id` and the value from `info.pokemon_id` in `quest_rewards` as the same values:
```
![image](https://user-images.githubusercontent.com/1327440/94358177-fb19f500-0053-11eb-9eab-ae040775644b.png)

```

**Comment by clburlison** [comment]
Yeah just verified with latest version of vanilla RDM. `quest_pokemon_id` is reserved for quests that give a Pokémon to catch. Initially this was getting populated to that column but later got removed. It seems Dataparser never had that logic updated after Mega Energy quests got added.

File: src/data/map.js [hunk 3]
```diff
@@ -481,6 +486,22 @@ const getPokestops = async (minLat, maxLat, minLon, maxLon, updated = 0, showPok
             excludePokemonSQL = sqlExcludeCreate;
```
Comment (Mygod) [inline src/data/map.js:486]: Either remove global energy filter and change line 447 and 449, or make the global filter work (and possibly add global Pokemon encounter reward filter too?).

**Review by Mygod — CHANGES_REQUESTED** [review]

**Pull request merged**